apps/framework-cli/src/mcp/tools/infra_issues/mod.rs

```diff
@@ -0,0 +1,849 @@
+//! # Infrastructure Issues Diagnostic Tool
+//!
+//! This module implements a proactive infrastructure diagnostics MCP tool that intelligently
+//! surfaces errors by using the infrastructure map to determine what to check based on
+//! infrastructure type (ClickHouse, Kafka/Redpanda, Temporal, etc.).
+//!
+//! Initial implementation focuses on ClickHouse diagnostics with extensible architecture
+//! for future infrastructure types.
+//!
+//! ## ClickHouse Diagnostic Providers
+//!
+//! The tool automatically runs multiple diagnostic providers based on table engine types:
+//!
+//! ### 1. MutationDiagnostic
+//! Detects stuck or failing mutations (ALTER operations) that can block table maintenance.
+//! - **Source**: `system.mutations`
+//! - **Detection**: Mutations not done (is_done = 0) or with non-empty failure reasons
+//! - **Thresholds**:
+//!   - Error: Mutation has a failure reason (latest_fail_reason not empty)
+//!   - Warning: Mutation in progress but not completed (is_done = 0)
+//! - **Suggested Action**: Cancel stuck mutations with KILL MUTATION
+//!
+//! ### 2. PartsDiagnostic
+//! Identifies excessive data parts per partition that impact query performance.
+//! - **Source**: `system.parts`
+//! - **Detection**: Active parts count per partition > 100
+//! - **Thresholds**:
+//!   - Error: part_count > 300
+//!   - Warning: 100 < part_count ≤ 300
+//! - **Suggested Action**: Run OPTIMIZE TABLE to merge parts
+//!
+//! ### 3. MergeDiagnostic
+//! Monitors long-running background merges.
+//! - **Source**: `system.merges`
+//! - **Detection**: Merges running > 300 seconds
+//! - **Thresholds**:
+//!   - Error: elapsed_time > 1800s (30 minutes)
+//!   - Warning: 300s < elapsed_time ≤ 1800s
+//! - **Note**: Progress is tracked and reported but not used in severity determination
+//! - **Suggested Action**: Monitor merge progress and check server resources (CPU, disk I/O, memory)
+//!
+//! ### 4. ErrorStatsDiagnostic
+//! Aggregates errors from ClickHouse system.errors to surface recurring issues.
+//! - **Source**: `system.errors`
+//! - **Detection**: All errors with count > 0 (reports top 10 by occurrence)
+//! - **Thresholds**:
+//!   - Error: error_count > 100
+//!   - Warning: error_count > 10
+//!   - Info: 0 < error_count ≤ 10
+//! - **Suggested Action**: Review error messages and recent system changes
+//!
+//! ### 5. S3QueueDiagnostic (S3Queue tables only)
```

Comment:
We probably want this for s3cluster as well? And/or drop s3 queue all together?

Reply:
s3Cluster is a synchronous function, it doesn't have the same caveats.
The S3 Table engine probably has some similar caveats alhough I don't think it will do anything untill you use it in a query so the feedback is actually synchronous there.